core/src/main/java/org/springframework/security/access/intercept/aopalliance/AuthorizationMethodInterceptor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.access.intercept.aopalliance;
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+
+import org.springframework.security.access.method.AuthorizationManagerAfterAdvice;
+import org.springframework.security.access.method.AuthorizationManagerBeforeAdvice;
+import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+/**
+ * Provides security interception of AOP Alliance based method invocations.
+ *
+ * @author Evgeniy Cheban
+ */
+public final class AuthorizationMethodInterceptor implements MethodInterceptor {
+
+	private final AuthorizationManagerBeforeAdvice<MethodInvocation> beforeAdvice;
+
+	private final AuthorizationManagerAfterAdvice<MethodInvocation> afterAdvice;
+
+	/**
+	 * Creates an instance.
+	 * @param beforeAdvice the {@link AuthorizationManagerBeforeAdvice} to use
+	 * @param afterAdvice the {@link AuthorizationManagerAfterAdvice} to use
+	 */
+	public AuthorizationMethodInterceptor(AuthorizationManagerBeforeAdvice<MethodInvocation> beforeAdvice,
```

Comment:
@rwinch made a good point that it would be nice to see if we can change `AuthorizationManagerBeforeAdvice` so that it doesn't extend `AuthorizationManager`.

Two ideas come to mind:

1. Change this constructor to
```java
public AuthorizationMethodInterceptor(AuthorizationManagerBeforeAdvice<MethodInvocation> beforeAdvice,
        AuthorizationManager<MethodInvocation> beforeManager,
        AuthorizationManagerAfterAdvice<MethodInvocation> afterAdvice,
        AuthorizationManager<MethodInvocation> afterManager)
```
and then `PreAnnotationAuthorizationManagerBeforeAdvice` would split into two classes --`PreFilterAuthorizationManagerBeforeAdvice` and `PreAuthorizeAuthorizationManager` -- and likewise for `PostAnnotationXXX`
2. Keep the contract as-is, but split `PreAnnotationAuthorizationManagerBeforeAdvice` into `PreFilterAuthorizationManagerBeforeAdvice` and something like `MethodMatcherAuthorizationManagerBeforeAdvice`. The latter would could have an instance of `AuthorizationManager` (e.g. `PreAuthorizeAuthorizationManager`) that it would delegate to if the `MethodMatcher` matches.

The second seems ideal, but I'm listing both for completeness.

In both cases, it seems like the before advice contract would change to return `void` since not all advices would return an `AuthorizationDecision` anymore.